### PR TITLE
Automated UI testing: project creation #495

### LINKED ISSUE #495: Automated UI test for the project creation
Section 2 in TESTING.md

### PR DESCRIPTION
**Pull request by OlliKiljunen**
# Description of the PR

Closes #495.

Kotlin is here! This PR contains initial steps for automated end-to-end testing.

That is, creating a new project and setting SDK to OpenJDK 11. Currently, the test only works if
run on clean IDEA.

For UI tests, a new source set "e2e" is introduced. IntelliJ IDEA does not automatically recognize the source set's code as "test" code (green color). For the best development experience, you should change that after updating Gradle (see pic).

screenshot

(Note: building, testing and other important things should work perfectly without any manual configuring. The only things that the fix above concerns are such as when you "Find Usages" in IDEA, whether the code in e2e source set are considered "production" code or "test" code. This is a bug in Gradle's IDEA plug-in that [apparently](https://youtrack.jetbrains.com/issue/IDEA-151925#focus=Comments-27-4614854.0-0) comes and goes, and JetBrains doesn't consider it that important.)

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [ ] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [ ] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -1,5 +1,6 @@
 plugins {
     id 'scala'
+    id 'org.jetbrains.kotlin.jvm' version '1.3.72'
```
Comment (OlliKiljunen) [inline build.gradle:3]: Newer Kotlin versions would require updating Gradle

File: build.gradle [hunk 2]
```diff
@@ -60,6 +83,11 @@ scalastyle {
     }
 }
 
+task uiTest(type: Test) {
+    testClassesDirs = sourceSets.e2e.output.classesDirs
+    classpath = sourceSets.e2e.runtimeClasspath
+}
```
Comment (OlliKiljunen) [inline build.gradle:89]: No other task currently depends on this. Trigger it manually after you have run **gradle clean** and **gradle runIdeForUiTests** is running.

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  fun simpleTest() = uiTest {
+    with(CommonSteps(this)) {
+      createProject()
+      closeTipOfTheDay()
+      downloadAndSetOpenJdk11()
+    }
+    with(ideFrame()) {
+      assertNotNull(menu().select("A+"))
+    }
+  }
```
Comment (OlliKiljunen) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:25]: This can be replaced with the real test procedure(s).

**Comment by OlliKiljunen** [comment]
@stellatsv I added you as a reviewer because I remembered you had learned some Kotlin. :)

File: build.gradle (hunk 1 above)
Comment (superseacat) [reply build.gradle:3]: `'1.4.30'` (latest) I wonder how large is this difference actually, is it like Scala's 2.12 vs 2.13 or else?

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/fixtures/CommonFixtures.kt [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+fun SearchContext.comboBox(name: String) = find(ComboBox::class.java,
+    byAttributeAndSwingClass("accessiblename", name, JComboBox::class.java))
+
+fun SearchContext.dialog(title: String, timeout : Duration = Duration.ofSeconds(5)) =
+    find(Dialog::class.java, byAttributeAndSwingClass("title", title, JDialog::class.java), timeout)
+
+fun SearchContext.heavyWeightWindow() = find(HeavyWeightWindow::class.java)
+
+@FixtureName("Welcome Frame")
```
Comment (superseacat) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/fixtures/CommonFixtures.kt:37]: custom DSL here were come! 🥳

**Comment by superseacat** [comment]
> IntelliJ IDEA does not automatically recognize the source set's code as "test" code (green color). For the best development experience, you should change that after updating Gradle (see pic).

for me IDEA did recognize the `e2e` as `sources root` due to being located in `src`, I'd actually suggest the following structure:

**Comment by superseacat** [comment]
I would like to know if the setup could work in a headless env (like GH action build) 🤓

File: build.gradle (hunk 2 above)
Comment (superseacat) [reply build.gradle:89]: can our custom **task** incorporate **gradle clean & gradle runIdeForUiTests**?

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import org.junit.Assert.*
+import org.junit.Test
+
+class PlaceholderTest {
+  init {
+    StepLoggerInitializer.init()
+  }
+
+  @Test
+  fun simpleTest() = uiTest {
```
Comment (superseacat) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:16]: fails for me, but it's clear how to fix it (probably a matter of time outing or splitting properly)

**Review by superseacat — APPROVED** [review]
almost forgot... great work! **A+** 😉

**Comment by superseacat** [comment]
> I would like to know if the setup could work in a headless env (like GH action build)

https://github.com/JetBrains/intellij-ui-test-robot/blob/master/ui-test-example/FAQ.md#can-i-run-the-tests-on-github-actions

File: build.gradle (hunk 1 above)
Comment (OlliKiljunen) [reply build.gradle:3]: It doesn't look like a big difference. They seem to be adding new versions very often (1.3.72 is still less than a year old).

File: build.gradle (hunk 2 above)
Comment (OlliKiljunen) [reply build.gradle:89]: Not easily, I'm afraid. The issue is, **runIdeForUiTests** should be run in parallel with **uiTest** and **clean** before **runIdeForUiTests**. AFAIK, Gradle doesn't support parallel tasks. The way they suggest running it [here](https://github.com/JetBrains/intellij-ui-test-robot#quick-start) is launching each task separately.

**Comment by OlliKiljunen** [comment]
> > IntelliJ IDEA does not automatically recognize the source set's code as "test" code (green color). For the best development experience, you should change that after updating Gradle (see pic).
> 
> for me IDEA did recognize the `e2e` as `sources root` due to being located in `src`,

To avoid any confusion, let's be precise: IDEA should (and at least it did for me) recognize `src/e2e/kotlin` as sources root. That is not, however, because it is located in `src` (`src/test/java` is too!) but because of these lines in `build.gradle`:

```
sourceSets {
    e2e {
        kotlin.srcDir "$projectDir/src/e2e/kotlin"
    }
}
```

IDEA simply assumes an included dir is sources root instead of test root (let's be honest, the opposite would be worse).

> I'd actually suggest the following structure:

That's how I organized it in the #492. In this time, I thought it would be beneficial to have UI tests in a separate source set (that does not see the production code but handles the plugin UI as a black box). It's apparent, both the approaches have their pros and cons (e.g. I dislike #492's ugly test filters in **build.gradle**).

But if we keep `e2e` as a separate source set (i.e. module), **how could we tell IDEA that `src/e2e/kotlin` should be test root instead of sources root?**

[Here's](https://youtrack.jetbrains.com/issue/IDEA-151925#focus=Comments-27-2355076.0-0) the answer. Basically, we add `idea` plugin to Gradle and configure:
```
idea {
  module {
    testSourceDirs += project.sourceSets.e2e.kotlin.srcDirs
  }
}
```
Too bad, that [doesn't work](https://youtrack.jetbrains.com/issue/IDEA-151925#focus=Comments-27-4614854.0-0) anymore with the newest IDEA versions (because of a bug?).

As the "fix" above were never meant to be the final solution to the issue, it is possible they're not paying much attention to get it working again. In the future, I hope there are gonna be proper ways to equip Gradle's source sets with meaningful IDEA semantics.

**Comment by OlliKiljunen** [comment]
> 
> 
> > I would like to know if the setup could work in a headless env (like GH action build)
> 
> https://github.com/JetBrains/intellij-ui-test-robot/blob/master/ui-test-example/FAQ.md#can-i-run-the-tests-on-github-actions

I suggest we create a new ticket of that because UI tests can be put into operation also even if they were still triggered manually.

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt (hunk 5 above)
Comment (OlliKiljunen) [reply src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:16]: If you paste the console output, I can tweak timeouts.

File: build.gradle (hunk 2 above)
Comment (superseacat) [reply build.gradle:89]: yes, despite doing that myself I've missed the async part, I'm curious to see how it behaves in terms of `runIdeFor...` termination. guess I want that to-be-created "ui-tests-in-CI" task for myself 😉

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt (hunk 5 above)
Comment (superseacat) [reply src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:16]: I will just fork and fix, no worries, but it's curious if it's the OS difference that makes it fail for me

File: build.gradle (hunk 2 above)
Comment (superseacat) [reply build.gradle:89]: this one https://github.com/Aalto-LeTech/intellij-plugin/issues/517 @OlliKiljunen

**Comment by superseacat** [comment]
regardless of where we put koltin, shall there be a checkstyle for it? @OlliKiljunen

**Comment by OlliKiljunen** [comment]
> 
> 
> regardless of where we put koltin, shall there be a checkstyle for it? @OlliKiljunen

I guess I should check what happens if I click this notification:

image

**Review by stellatsv — APPROVED** [review]
The test failed for me as well but I see there is plan to fix that. Other than that, it worked fine locally.

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt (hunk 5 above)
Comment (superseacat) [reply src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:16]: on the other hand, I see no problem if you fix it yourself 😛

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package fi.aalto.cs.apluscourses.e2e.utils
+
+import com.intellij.remoterobot.RemoteRobot
+import com.intellij.remoterobot.stepsProcessing.StepLogger
+import com.intellij.remoterobot.stepsProcessing.StepWorker
+
+object StepLoggerInitializer {
+  private var initialized = false;
```
Comment (superseacat) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt:8]: gotcha, you, nasty Java developer! :wink:

File: build.gradle (hunk 1 above)
Comment (nikke234) [reply build.gradle:3]: Should we update?

Comment (nikke234) [reply build.gradle:3]: But anyways that can be done separately from this PR

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.intellij.remoterobot.RemoteRobot
+import com.intellij.remoterobot.stepsProcessing.StepLogger
+import com.intellij.remoterobot.stepsProcessing.StepWorker
+import java.time.Duration
+
+object StepLoggerInitializer {
+  private var initialized = false
+  fun init() = synchronized(initialized) {
+    if (!initialized) {
```
Comment (nikke234) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt:11]: Could use an atomic boolean with
```
if (initialized.compareAndSet(false, true)) { ... }
```

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package fi.aalto.cs.apluscourses.e2e.utils
+
+import com.intellij.remoterobot.RemoteRobot
+import com.intellij.remoterobot.stepsProcessing.StepLogger
+import com.intellij.remoterobot.stepsProcessing.StepWorker
+import java.time.Duration
+
+object StepLoggerInitializer {
```
Comment (nikke234) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt:8]: I don't know much about Kotlin, but doesn't it allot top level functions and variables? If so, can we remove the enclosing object?

**Review by nikke234 — COMMENTED** [review]
Very cool stuff and nice implementation! I can't get the tests to succeed at the moment, but if I remember correctly that was a known issue?

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package fi.aalto.cs.apluscourses.e2e
+
+import fi.aalto.cs.apluscourses.e2e.fixtures.*
+import fi.aalto.cs.apluscourses.e2e.steps.CommonSteps
+import fi.aalto.cs.apluscourses.e2e.utils.StepLoggerInitializer
+import fi.aalto.cs.apluscourses.e2e.utils.uiTest
+import org.junit.Assert.*
+import org.junit.Test
+
+class PlaceholderTest {
```
Comment (nikke234) [inline src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:10]: Just making sure I understand, is this like a wrapper around the UI tests that wraps it in a normal JUnit test?

**Comment by nikke234** [comment]
Tests pass now for me as well :tada: We might want to consider removing the JDK-installation from the UI tests, but it's also fine as is :+1:

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt (hunk 7 above)
Comment (OlliKiljunen) [reply src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt:11]: I'll change that.

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt (hunk 8 above)
Comment (OlliKiljunen) [reply src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/utils/Utils.kt:8]: My implementation is actually a crude adaptation of [the example project's logger](https://github.com/JetBrains/intellij-ui-test-robot/blob/master/ui-test-example/src/test/kotlin/org/intellij/examples/simple/plugin/utils/StepsLogger.kt). Their reason to encapsulate it inside an object is probably making sure no one else touches `initialized` field. I think it's fair enough to keep it that way.

File: src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt (hunk 9 above)
Comment (OlliKiljunen) [reply src/e2e/kotlin/fi/aalto/cs/apluscourses/e2e/PlaceholderTest.kt:10]: There are very little "wrapping" actually.

If you mean these lines:
```
fun uiTest(test: RemoteRobot.() -> Unit) {
  RemoteRobot("http://localhost:8082").test()
}
```
then yeah, they kinda "wrap" the method but basically they just create a `RemoteRobot` object and available via `this` keyword.

So basically our UI tests are nothing more than JUnit tests that starts with a RemoteRobot object initialization. The remote robot connects to the (localhosted) server put up by `runIdeForUiTests` and does all the mouse-moving magic. Our JUnit test(s) just call methods of that robot and asserts it returns correct values.

**Pull request merged**